### PR TITLE
8266440: Shenandoah: TestReferenceShortcutCycle.java test failed on AArch64

### PR DESCRIPTION
**Pull request by zhengyu123**
JDK-8263427 missed corresponding changes for aarch64.

LRB needs to be enabled during weak roots/references processing, regardless if heap is stable.

Test:

- [x] hotspot_gc_shenandoah on Macosx/AArch64
- [ ] hotspot_gc_shenandoah on Linux 64/AArch64

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8266440](https://bugs.openjdk.java.net/browse/JDK-8266440): Shenandoah: TestReferenceShortcutCycle.java test failed on AArch64

### Reviewers
 * [Andrew Haley](https://openjdk.java.net/census#aph) (@theRealAph - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3902/head:pull/3902` \
`$ git checkout pull/3902`

Update a local copy of the PR: \
`$ git checkout pull/3902` \
`$ git pull https://git.openjdk.java.net/jdk pull/3902/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3902`

View PR using the GUI difftool: \
`$ git pr show -t 3902`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3902.diff">https://git.openjdk.java.net/jdk/pull/3902.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/aarch64/gc/shenandoah/shenandoahBarrierSetAssembler_aarch64.cpp [hunk 1]
```diff
@@ -242,7 +242,21 @@ void ShenandoahBarrierSetAssembler::load_reference_barrier(MacroAssembler* masm,
   __ ldrb(rscratch2, gc_state);
 
   // Check for heap stability
-  __ tbz(rscratch2, ShenandoahHeap::HAS_FORWARDED_BITPOS, heap_stable);
+  if (is_strong) {
+    __ tbz(rscratch2, ShenandoahHeap::HAS_FORWARDED_BITPOS, heap_stable);
+  } else {
+    if (dst == rscratch1) {
+      __ push(rscratch1, sp);
```
Comment (nick-arm) [inline src/hotspot/cpu/aarch64/gc/shenandoah/shenandoahBarrierSetAssembler_aarch64.cpp:249]: Could we avoid the push/pop by passing in an extra temporary? We have `tmp1` available in `ShenandoahBarrierSetAssembler::load_at()` which is the only place that calls this.

Comment (zhengyu123) [reply src/hotspot/cpu/aarch64/gc/shenandoah/shenandoahBarrierSetAssembler_aarch64.cpp:249]: Unfortunately, tmp1 == noreg here.

Comment (zhengyu123) [reply src/hotspot/cpu/aarch64/gc/shenandoah/shenandoahBarrierSetAssembler_aarch64.cpp:249]: We need to test 2 bits here.

Comment (zhengyu123) [reply src/hotspot/cpu/aarch64/gc/shenandoah/shenandoahBarrierSetAssembler_aarch64.cpp:249]: so, this is better alternative?
    Label lrb;
    __ tbnz(rscratch2, ShenandoahHeap::HAS_FORWARDED_BITPOS, lrb);
    __ tbnz(rscratch2, ShenandoahHeap::WEAK_ROOTS_BITPOS, lrb);
    __ b(heap_stable);

    __ bind(lrb);

Comment (theRealAph) [reply src/hotspot/cpu/aarch64/gc/shenandoah/shenandoahBarrierSetAssembler_aarch64.cpp:249]: How about
```
 // Check for heap stability                                                                                                                                                                
  if (is_strong) {
    __ tbz(rscratch2, ShenandoahHeap::HAS_FORWARDED_BITPOS, heap_stable);
  } else {
    label lrb;
    __ tbnz(rscratch2, ShenandoahHeap::WEAK_ROOTS_BITPOS, lrb);
    __ tbz(rscratch2, ShenandoahHeap::HAS_FORWARDED_BITPOS, heap_stable);
    __ bind(lrb);
  }
```

Comment (zhengyu123) [reply src/hotspot/cpu/aarch64/gc/shenandoah/shenandoahBarrierSetAssembler_aarch64.cpp]: Yep, much better. Thanks.

**Review by theRealAph — APPROVED** [review]

**Pull request closed without merging**